File: implementaciones_examen.hs
{-
Hice esto con compañeros
-}

cambiar :: [(Char,Char)] -> [Char] -> [Char]
cambiar _ [] = []
cambiar diccionario (x:xs) = cambiarCaracter diccionario x : cambiar diccionario xs

cambiarCaracter :: [(Char,Char)] -> Char -> Char
cambiarCaracter ((x,y):xs) caracter | caracter == x = y
                                    | otherwise = cambiarCaracter xs caracter

verifica :: [(Char,Char)] -> [[Char]] -> [[Char]] -> Bool
verifica _ [] [] = True
verifica b (m:ms) (n:ns) | cambiar b m == n = True && verifica b ms ns


secuencia :: [Integer] -> Integer -> [Integer]
secuencia [] _ = []
secuencia (x:xs) n | x `mod` 2 == 0 = (x*n) : (secuencia xs n)
                   | otherwise = x : (secuencia xs n)

borrarMult3 :: [Integer] -> [Integer]
borrarMult3 [] = []
borrarMult3 (x:xs) | x `mod` 3 == 0 = borrarMult3 xs
                   | otherwise = x : borrarMult3 xs

esMatriz :: [[Integer]] -> Bool
esMatriz [] = True 
esMatriz (x:y:xs) | longitud x == longitud y = True && esMatriz xs
                  | otherwise = False

longitud :: [t] -> Integer
longitud [] = 0
longitud (x:xs) = 1 + longitud xs

pertenece :: (Eq t) => t -> [t] -> Bool
pertenece _ [] = False
pertenece e (x:xs) = e == x || pertenece e xs

perteneceA :: [[Integer]] -> Integer -> [[Bool]]
perteneceA [] _ = []
perteneceA (x:xs) a = [pertenece a x] : (perteneceA xs a)

fo :: [[Integer]] -> [Bool]
fo [] = []
fo (x:xs) | ordenados x = True : fo xs
          | otherwise = False : fo xs

ordenados :: [Integer] -> Bool
ordenados [] = True
ordenados [e] = True
ordenados (x:y:xs) | (x <= y) = ordenados (y:xs)
                   | otherwise = False

n1 :: [Integer] -> Integer -> Integer
n1 (x:xs) 0 = x
n1 (x:xs) n = n1 xs (n-1)

buscarPosicion :: [Integer] -> Integer -> Integer
buscarPosicion [] _ = undefined
buscarPosicion (x:xs) c | x == c = 0
                        | otherwise =  1 + buscarPosicion xs c

quitarRepetidos :: [Integer] -> [Integer]
quitarRepetidos [] = []
quitarRepetidos [x] = [x]
quitarRepetidos (x:y:xs) | x == y = x : quitarRepetidos xs
                         | otherwise = x : y : quitarRepetidos xs

File: implementaciones_examen.py
from typing import List, Tuple

def cambiar(diccionario: List[Tuple[str, str]], aCambiar: List[str]) -> List[str]:
    res: List[str] = [] # Inicializo lista vacía.
    for i in aCambiar: # Recorro los caracteres que quiero cambiar
        for j in diccionario: # recorro los valores "key" de mi diccionario
            if i == j[0]: # si i matchea con el valor "key" de mi diccionario.
                res.append(j[1]) # Metele el segundo valor "value" de mi diccionario.
    return res

#print(cambiar([('a', 'A'), ('b', 'B')], ['a', 'b']))

def verifica(b:List[Tuple[str,str]],m:List[List[str]],n:List[List[str]])->bool:
    for i in range(len(m)):
        if cambiar(b,m[i]) != n[i]:
            return False
    return True

#print(verifica([('a','A'),('b','B')], [['a'],['b']], [['A'],['B']]))


def secuencia (L:List[int], n:int)->List[int]:
    res : List[int] = [] # Inicializo lista vacía
    for i in range(len(L)):
        if L[i] % 2 == 0:
            res.append(L[i]*n)
        else:
            res.append(L[i])
    return res

#print(secuencia([1,2,3,4],0))




def pertenece(s:list,e:int)->bool:
    for i in s:
        if i == e:
            return True
    return False

def perteneceACadaUno(s:list,e:int)->list:
    res:list=[]
    for i in s:
        if pertenece(i,e):
            res.append(True)
        else:
            res.append(False)
    return res

#print(perteneceACadaUno([[1,2,3],[1,2,3]],3))

def ordenados(s:list)->bool:
    for i in range(len(s)-1):
        if s[i] >= s[i+1]:
            return False
    return True

def filasOrdenadas(m: List[List[int]]) -> List[bool]:
    res:list[int]=[]
    for i in m:
        if ordenados(i):
            res.append(True)
        else:
            res.append(False)
    return res

#print(filasOrdenadas([[0,1,2,3],[0,1,2,3],[5,2,22]]))

def n1 (L:List[int],n:int)->int:
    return L[n]

#print(n1([1,2,3,4],0))

def quitarRepetidos (b:List[int])->List[int]:
    res:list[int]=[]
    for i in b:
        if i not in res:
            res.append(i)
    return res

print(quitarRepetidos([1,0,1,0,2,6,8,2]))